### ISSUE TITLE
Optimise boolean expression within where argument

### ISSUE THREAD

**Issue by johngrimes** [aehrc/pathling#142]
The following expression is quite simple but performs exceptionally poorly. It also seems to use a very large amount of memory.

```
reverseResolve(Observation.subject)
    .where(
        $this.status = 'preliminary'
        or
        $this.status = 'final'
    ).status
```

**Comment by johngrimes** [comment]
Interestingly, the following request is very slow and fails:

```
AggregateRequest(subjectResource=PATIENT, aggregations=[AggregateRequest.Aggregation(label=Optional[Number of patients], expression=count())], groupings=[AggregateRequest.Grouping(label=Optional[Observation status], expression=reverseResolve(Observation.subject)
    .where(
        $this.status = 'preliminary'
        or
        $this.status = 'final'
    ).status)], filters=[])
```

While the following request is fine:

```
AggregateRequest(subjectResource=OBSERVATION, aggregations=[AggregateRequest.Aggregation(label=Optional[Number of observations], expression=count())], groupings=[AggregateRequest.Grouping(label=Optional[Code contains expression], expression=where(
    $this.status = 'preliminary'
    or
    $this.status = 'final'
).status)], filters=[])
```

**Comment by piotrszul** [comment]
I did not look at the details but off the top of my head one thing to consider is that without any extra logic skips joins if $this is used multiple times or caching of $this dataset, $this expression (including all the joins it contains with reverseResolve) will be evaluated multiple times. 
In the second example $this is actually cached (all resource datasets are).
So perhaps an easy fix would be the cache() $this?

**Comment by johngrimes** [comment]
I have done a bit of an experiment with inserting call to `Dataset.cache()` following the creation of the `FhirPath` instance representing the `$this` context (`InvocationVisitor#visitFunctionInvocation`).

Unfortunately it does not seem to improve performance, which is still quite poor. In fact, it seems to make it worse.

Without `Dataset.cache()`: ~ 20 sec
With `Dataset.cache()`: ~ 31 sec

**Comment by piotrszul** [comment]
Hmm. Interesting. Maybe spark has some way to internally recognize
duplicate joins.
Maybe comparing/looking at the query plans could shed some light on how to
make the implementation more efficient.

**Comment by piotrszul** [comment]
Perfomance testing results:
[Pathling-Performance.xlsx](https://github.com/aehrc/pathling/files/5388946/Pathling-Performance.xlsx)

**Comment by johngrimes** [comment]
I think that a more efficient way of writing this query might be:

```
reverseResolve(Observation.subject)
    .where($this.status in ('preliminary' | 'final'))
    .status
```

This would require the implementation of the [union operator](http://hl7.org/fhirpath/#unionother-collection).

**Comment by johngrimes** [comment]
We've identified a change that makes a big difference to the performance of this query, which is to change the representation of `ResourcePath` from a struct to a set of regular columns.

The former seems to get in the way of Spark's query optimisation, impeding its ability to selectively retrieve data from Parquet.

**Comment by johngrimes** [comment]
In changing the implementation to change the representation of `ResourcePath` (and reduce joins where possible), I've identified that changing the way we implement aggregate functions could simplify things (and improve performance).

The proposal is to move to windowing functions for this purpose, and I've created examples of this method and the current method for comparison: https://gist.github.com/johngrimes/425119410742e71b020e4f2ccd657f5d

**Comment by johngrimes** [comment]
Fixed by #191.
